Use server side generated timestamps for :relative-datetime

lbrdnk opened the pull request:
Closes #21518

### Description

This PR modifies transformation of `:relative-datetime` for redshift driver in a way, that it uses server side generated timestamp instead of `getdate()` call, for units coarser or eq to `:day`. This enables caching of queries that use `:relative-datetime`.

*What to modify.* To the best of my knowledge `:relative-datetime` is the only clause that uses honeysql's `:%now`, hence the only place where it makes sense to modify the behavior.

*Matching the `getdate` and server side timestamp values.* [Docs](https://docs.aws.amazon.com/redshift/latest/dg/r_GETDATE.html) state that `getdate`:
> ...returns the current date and time in the current session time zone... 

Sever side value is generated using `qp.timezone/now` which uses the results timezone which corresponds to session timezone in Redshift terms. To avoid jdbc driver's timezone adjustments, the value is sql string formatted and passed to database that way. On db side cast is used to re-create the timestamp.

*Further work.* Similar issue is reported for snowflake driver in #20988.

### How to verify

New tests were added:
1. Test checking that query using `:relative-datetime` with granularity greater than or equal to day (`:week` is used in the test) was actually cached.
2. Tests that check the timestamp generated using `:relative-datetime` with `getdate` under the hood is equal to timestamp generated using `:relative-datetime` using server side generated timestamp.
3. Test that checks that results of query asked in New York (report timezone) are correct if the system is in UTC and has already the next week.

### Checklist

- [X] Tests have been added/updated to cover changes in this PR

lbrdnk committed:
WIP: Use server side generated timestamps for :relative-datetime

lbrdnk committed:
Use correct timestamp in :relative-datetime and add tests

lbrdnk committed:
Update tests

lbrdnk committed:
Remove combinations of tzs test

I've added those to commit in case we
need to refer to it in future. Tests like that take ~500 seconds,
hence good for one off testing.

lbrdnk committed:
Update docstrings and comments

metamben commented on modules/drivers/redshift/test/metabase/driver/redshift_test.clj:
```diff
@@ -391,3 +394,117 @@
+   (getdate-vs-ss-ts-test-thunk-generator :week -1))
+  ([unit value]
+   (fn []
+     (let [honey {:select [[(with-redefs-fn {#'redshift/use-server-side-timestamp? (constantly false)}
+                              #(sql.qp/->honeysql :redshift [:relative-datetime value unit]))]
```
This looks suspicious. When does the lambda run? It seems to escape the scope of the `with-redefs-fn` and run with the restored settings. Either the lambda has to be removed or `with-redefs-fn` has to be moved inside it.

Are you playing with the lambda to make it less likely that the two expressions are executed on different days? If so, it might be easier to skip the test in the "danger zone".

lbrdnk replied:
I've modified the code to use just `with-redefs`, not the `with-redefs-fn` variant.

metamben commented on modules/drivers/redshift/test/metabase/driver/redshift_test.clj:
```diff
@@ -391,3 +394,117 @@
+(deftest ss-timestamp-test
+  (mt/test-driver
+   :redshift
+   (testing "Value of getdate() and server side generated timestamp is equal"
```
```suggestion
   (testing "Value of getdate() and server side generated timestamp are equal"
```

lbrdnk committed:
Use string formatted qp.timezone/now as a base

lbrdnk committed:
Merge remote-tracking branch 'origin/master' into redshift-getdate

lbrdnk commented on modules/drivers/redshift/test/metabase/driver/redshift_test.clj:
```diff
@@ -391,3 +394,117 @@
+                        :params params})
+      qp/process-query))
+
+(defn- session-id! []
```
I think I wanted to emphasize that session id is gathered querying the database ie. depends on the effects of the outside world. However with this logic I should probably end most of the titles of functions that I write with exclamation. Hence removing it.

lbrdnk commented on modules/drivers/redshift/test/metabase/driver/redshift_test.clj:
```diff
@@ -391,3 +394,117 @@
+                       "limit 5\n")]
+    (->> (run-native-query query-str)
+         mt/rows
+         (map (fn [[session-id start-time result-cache-hit query-text]]
+                {:session-id session-id
+                 :start-time start-time
+                 :result-cache-hit result-cache-hit
+                 :query-text query-text})))))
```
Thank you :)

lbrdnk commented on modules/drivers/redshift/test/metabase/driver/redshift_test.clj:
```diff
@@ -391,3 +394,117 @@
+(defn- last-queries-of-a-session [session-id]
+  (let [query-str (str "select session_id, start_time, result_cache_hit, query_text\n"
+                       "from sys_query_history\n"
+                       "where session_id = " session-id "\n"
```
While writing that I was thinking that here it does not matter much. However, reading your comment I have a feeling it could be a good habit using params by default. I've changed it to param.

lbrdnk commented on modules/drivers/redshift/test/metabase/driver/redshift_test.clj:
```diff
@@ -391,3 +394,117 @@
+      qp/process-query))
+
+(defn- session-id! []
+  (-> (run-native-query "select pg_backend_pid()") mt/rows ffirst))
```
Done.

lbrdnk commented on modules/drivers/redshift/test/metabase/driver/redshift_test.clj:
```diff
@@ -391,3 +394,117 @@
+  (let [query-str (str "select session_id, start_time, result_cache_hit, query_text\n"
+                       "from sys_query_history\n"
+                       "where session_id = " session-id "\n"
+                       "order by start_time desc\n"
```
My understanding here is, that as `qp/process-query` is synchronous function, it should not be necessary to check for that. Please correct me if I'm wrong.

metamben replied:
If the resolution of `start_time` is not fine granular enough, then two quick queries can get the same value. If that happens, ordering by `start_time` doesn't help.

lbrdnk replied:
Understood. I did as proposed. Comment linking this discussion was added.

lbrdnk commented on modules/drivers/redshift/test/metabase/driver/redshift_test.clj:
```diff
@@ -391,3 +394,117 @@
+            (testing "Last queries results contain cached query on expected position"
+              (is (.contains cached-row-adjusted-redshift-sql compiled-sql)))
+            (testing "Query was returned from cache"
+              (is (true? (:result-cache-hit cached-row)))))))))))
```
I'd propose first to add a comment with the concern. Then, If we find that test flaking, I should either consider some other way of testing it or remove it completely.

However, [Redshift docs](https://docs.aws.amazon.com/redshift/latest/dg/c_challenges_achieving_high_performance_queries.html) say that "...caching is transparent..." and that Redshift "...uses cached results for a new query when [specific conditions] are true...".

metamben replied:
> "...caching is transparent..." and that Redshift "...uses cached results for a new query when [specific conditions] are true..."

Sounds like they are not promising anything. Do they define what those specific conditions are?

metamben replied:
Yeah, these are the necessary but not the sufficient conditions for caching.

lbrdnk committed:
Address review remarks

metamben commented on modules/drivers/redshift/test/metabase/driver/redshift_test.clj:
```diff
@@ -391,3 +395,146 @@
+                       "from sys_query_history\n"
+                       "where session_id = ? \n"
+                       "order by start_time desc\n"
+                       "limit 5\n")]
```
Why five? The test looks at three, doesn't it?

lbrdnk committed:
Remove redundant data from "last queries" query

metamben reviewed (approved):
There are just two small remarks left.

lbrdnk committed:
Update tests

- Add comment that points on discussion on flake potential of a test.
- Correct the order of expected and actual in another test.
- Add formatting to rows and select only necessary for testing in a test.

lbrdnk committed:
Merge remote-tracking branch 'origin/master' into redshift-getdate

lbrdnk commented on modules/drivers/redshift/test/metabase/driver/redshift_test.clj:
```diff
@@ -391,3 +395,146 @@
+                  [[13 "Dwight Gresham" "2014-08-01T10:30:00-04:00" "75a1ebf1-cae7-4a50-8743-32d97500f2cf"]
+                   [15 "Rüstem Hebel" "2014-08-01T12:45:00-04:00" "02ad6b15-54b0-4491-bf0f-d781b0a2c4f5"]
+                   [7 "Conchúr Tihomir" "2014-08-02T09:30:00-04:00" "900335ad-e03b-4259-abc7-76aac21cedca"]
+                   [6 "Shad Ferdynand" "2014-08-02T12:30:00-04:00" "d35c9d78-f9cf-4f52-b1cc-cb9078eebdcb"]]))))))))
```
Thanks. Order is now correct.
